drivers/spi/spi_gd32.c

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2021 BrainCo Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_spi
+
+#include <errno.h>
+#include <kernel.h>
+#include <drivers/pinctrl.h>
+#include <drivers/spi.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(spi_gd32);
+
+#include "spi_context.h"
+
+/* SPI error status mask */
+#define SPI_GD32_ERR_MASK	(SPI_STAT_RXORERR | SPI_STAT_CONFERR | SPI_STAT_CRCERR)
+
+struct spi_gd32_config {
+	uint32_t reg;
+	uint32_t rcu_periph_clock;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+struct spi_gd32_data {
+	struct spi_context ctx;
+};
+
+static int spi_gd32_get_err(uint32_t spi)
+{
+	uint32_t stat = SPI_STAT(spi);
+
+	if (stat & SPI_GD32_ERR_MASK) {
+		LOG_WRN("spi%u error status detected, err = %u",
+			spi, stat & (uint32_t)SPI_GD32_ERR_MASK);
+
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static bool spi_gd32_transfer_ongoing(struct spi_gd32_data *data)
+{
+	return spi_context_tx_on(&data->ctx) ||
+	       spi_context_rx_on(&data->ctx);
+}
+
+static int spi_gd32_configure(const struct device *dev,
+			      const struct spi_config *config)
+{
+	struct spi_gd32_data *data = dev->data;
+	const struct spi_gd32_config *cfg = dev->config;
+	spi_parameter_struct spi_parameter;
+
+	if (spi_context_configured(&data->ctx, config)) {
+		return 0;
+	}
+
+	if (SPI_OP_MODE_GET(config->operation) == SPI_OP_MODE_SLAVE) {
+		LOG_ERR("Slave mode not support");
+		return -ENOTSUP;
+	}
+
+	spi_disable(cfg->reg);
+
+	spi_parameter.device_mode = SPI_MASTER;
+	spi_parameter.trans_mode = SPI_TRANSMODE_FULLDUPLEX;
+
+	if (SPI_WORD_SIZE_GET(config->operation) == 8) {
+		spi_parameter.frame_size = SPI_FRAMESIZE_8BIT;
+	} else {
+		spi_parameter.frame_size = SPI_FRAMESIZE_16BIT;
+	}
+
+	if (config->cs) {
+		spi_parameter.nss = SPI_NSS_SOFT;
+	} else {
+		spi_parameter.nss = SPI_NSS_HARD;
+		spi_nss_output_enable(cfg->reg);
+	}
+
+	if (config->operation & SPI_TRANSFER_LSB) {
+		spi_parameter.endian = SPI_ENDIAN_LSB;
+	} else {
+		spi_parameter.endian = SPI_ENDIAN_MSB;
+	}
+
+	spi_parameter.clock_polarity_phase = 0U;
+
+	if (config->operation & SPI_MODE_CPOL) {
+		spi_parameter.clock_polarity_phase |= SPI_CTL0_CKPL;
+	}
+
+	if (config->operation & SPI_MODE_CPHA) {
+		spi_parameter.clock_polarity_phase |= SPI_CTL0_CKPH;
+	}
+
+	/* Default spi clock prescaler to PCLK/32 */
+	spi_parameter.prescale = SPI_PSC_32;
+
+	spi_init(cfg->reg, &spi_parameter);
```

Comment:
non-blocking: consider using lower level HAL register access (it's confusing to understand what is HAL vs Zephyr code here)

Reply:
That's a good ideal, I will try to use the raw register operator to implement it.

Reply:
done.